### PR TITLE
Pun Pun has gained a few persistence save file features

### PR DESCRIPTION
**Pull request by Incoming5643**
Pun Pun has gained a few persistence save file features
- Whatever Pun Pun is wearing on his/her head and face at round end will still be there at the start of the next round. Back and hands do not save for obvious balance reasons.
- If Pun Pun dies but leaves an intact corpses his/her successor will be named Pun Pun II and inherit the items his father's/mother's corpse was wearing at the end of the round.
- If Pun Pun gets gibbed the chain of monkeys is broken and the next round monkey will be a new progenitor with a potentially different name. Obviously their items won't be saved in this case.

**Suggest as many monkey names as you want**

https://tgstation13.org/phpBB/viewtopic.php?f=10&t=5917 <-- Monkey name forum suggestion thread

### CODE REVIEW

**Comment by as334** [comment]
Wordplay Wordplay

File: code/modules/mob/living/carbon/monkey/punpun.dm [hunk 1]
```diff
@@ -0,0 +1,65 @@
+/mob/living/carbon/monkey/punpun //except for a few special persistence features, pun pun is just a normal monkey
+	name = "Pun Pun" //C A N O N
+	unique_name = 0
+	var/ancestor_name
+	var/ancestor_chain = 1
+	var/relic_hat	//Note: these two are paths
+	var/relic_mask
+	var/memory_saved = 0
+	var/list/pet_monkey_names = list("Pun Pun", "Bubbles", "Mojo", "Forever")
+
```
Comment (KorPhaeron) [inline code/modules/mob/living/carbon/monkey/punpun.dm]: Where is Mr Deempisi

**Comment by Incoming5643** [comment]
caveats: This makes punpun a subtype of normal monkeys, meaning there's a few ways to basically "erase" his uniqueness by turning him into a human and then a normal monkey. Solving this is not a priority as his save file will be fine even if he gets genetic'd out of his role.

Also it will require a map edit, which I will do at the very END of this pull for the sake of my sanity

File: code/modules/mob/living/carbon/monkey/punpun.dm (hunk 1 above)
Comment (Incoming5643) [inline code/modules/mob/living/carbon/monkey/punpun.dm]: There's a little portrait of Mr Deempisi saying he was killed in the bar, I'm sorry his death is canon.

But yeah throw ape names at me.

**Comment by MMMiracles** [comment]
Duplicate all the swat helms/voice changers

File: code/modules/mob/living/carbon/monkey/punpun.dm [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+	var/savefile/S = new /savefile("data/npc_saves/Punpun.sav")
+	if(gibbed)
+		S["ancestor_name"] 		<< null
+		S["ancestor_chain"]		<< 1
+		S["relic_hat"]			<< null
+		S["relic_mask"]			<< null
+		return
+	if(dead)
+		S["ancestor_name"] 		<< ancestor_name
+		S["ancestor_chain"]		<< ancestor_chain + 1
```
Comment (RemieRichards) [inline code/modules/mob/living/carbon/monkey/punpun.dm]: ```++ancestor_chain``` pls.

Comment (Incoming5643) [inline code/modules/mob/living/carbon/monkey/punpun.dm]: I have a reason for this one: Write memory is called twice for some dead punpuns: Once when he dies, and once at the end of the round/later gibbing. If I use `++ancestor_chain` I'm pretty sure I actually increment the var, which means that Pun Pun II could be followed by Pun Pun IV

Comment (RemieRichards) [inline code/modules/mob/living/carbon/monkey/punpun.dm]: sounds like a flawed system tbh but k.

Comment (duncathan) [inline code/modules/mob/living/carbon/monkey/punpun.dm]: Pun Pun III was the short lived second son of Pun Pun

Comment (Incoming5643) [inline code/modules/mob/living/carbon/monkey/punpun.dm]: It's also a safety against 
* badmin shinanagans of spawning multiple pun puns
* Healing wands

I keep the mob vars from changing under any circumstance. so even if they're called multiple times somehow they only ever increment once

Comment (YotaXP) [inline code/modules/mob/living/carbon/monkey/punpun.dm]: Incoming is doing it correctly.  `++` is the alternative to `+=1`.  This is not a `+=1`

File: code/modules/mob/living/carbon/monkey/punpun.dm (hunk 1 above)
Comment (YotaXP) [inline code/modules/mob/living/carbon/monkey/punpun.dm]: Could make Mr Deempisi start off with `ancestor_chain = 2`

**Comment by PKPenguin321** [comment]
Just a thought, maybe make hats with any armor values not get duplicated.

**Comment by Lumipharon** [comment]
This is neat, but of course some memester will, as mentioned above, put on HoS hats/voice changers on them for ebin duplication.

Not the biggest balance concern, but it is there.

**Comment by KorPhaeron** [comment]
Though a very far fetched/rare scenario, the ninja putting his antag hunter mask on the monkey to use next round might suck

**Comment by Shadowlight213** [comment]
@KorPhaeron ninja antag hud doesn't work. Hasn't worked in ages. Their mask is just a fancy voice changer currently.

**Comment by KazeEspada** [comment]
speaking of which, voice change gas masks.

**Comment by Iamgoofball** [comment]
honestly if you can put a voice changer on Pun Pun and have it not get stolen, kudos to you

**Comment by Incoming5643** [comment]
There's also the very real chance of not getting back whatever you put on pun pun next round before someone else. Especially if you don't manage to roll Bartender/Clown/Mime/That one assistant spawn in the bar.

**Comment by Steelpoint** [comment]
Restrict certain items like the HoS Hat or antag headware for appearing in the next round, for balance reasons. We already have enough problems with free Nuke Op gear coming in every round.

**Comment by PKPenguin321** [comment]
or do what i said and check if the hat's armor values are 0, and if they're not don't duplicate them

**Comment by Incoming5643** [comment]
Blacklists are bad. Fix the away missions having hundreds of telecrystals in free shit.

But yeah that's my go to on this one, the fact that there's almost no assurances of personally recovering whatever you put on pun pun, ESPECIALLY if it's something people know is good.

**Comment by octareenroon91** [comment]
>Blacklists

If you babysit Pun Pun all round, you can send a hat and a mask to the next shift. While they certainly could be disruptive to the early game, how likely is it that they will unfairly tip the balance?

Give him a 'collectible' version of the same hat?

**Comment by KorPhaeron** [comment]
Alright if the ninja mask is broken then this PR is completely fine balance/design whatever wise.

Blacklist is not needed, don't get worked up over a single armoured hat being duped.

**Comment by PKPenguin321** [comment]
my suggestion isn't a blacklist, it's just changing "if(pun pun is wearing a hat)" to "if(pun pun is wearing a hat && hat.armor == 0)" or whatever (this obviously isn't how the exact code would look but you know what i mean)

**Comment by Shadowlight213** [comment]
@KorPhaeron ninja hood prevents ai tracking similar to an agent id.

@incoming5643 away mission balance is completely unrelated to this. The gateway can be turned off with a simple config change.
I'm mainly worried about an admin running an event and spawning some super strong badmin only headgear and some player plopping it on punpun.
Next round, the admin logs off and suddenly you have Uber helmet of invincibility available for the crew. Keep the name thing, but having the head and mask be stored is going too far.

**Comment by Incoming5643** [comment]
var editing isn't preserved in this process, so anything edited to be overpowered wouldn't still be overpowered come the next round.

About the most obnoxious think pun pun could stow away to the next round that is adminspawn only is a helmet of justice

**Comment by Shadowlight213** [comment]
I was thinking more along the lines of something like an ert helmet, but I guess they are all hardsuits now. I swear there was something that had ert helmet armor but wasn't a hardsuit.
Also justice helmets aren't adminspawn only. They can be gotten from cargo.

**Comment by Incoming5643** [comment]
Alright for sake of argument I combed through all the head/mask items in the game, here are the ones that provide some special benefit above and beyond what is normally available to spessmen.

* "Easter Bunny Head": Gives -1 runspeed (easter only)
* "Centcom hat": Actually has worse armoring than the HoS Hats
* "drone (hiding)": Doesn't actually contain a drone in this instance, is just a hat
* "agent headgear": Potential loot from an abductor, but no special features
* "Prophet's Hat": Potential loot from hand of god, has really excellent armoring (melee = 60, bullet = 60, laser = 60, energy = 50, bomb = 70, bio = 50, rad = 50), but hand of god is kind of dead.
* "Chronosuit Helmet": Potential loot from a chronolegionnaire, pretty crazy armoring (melee = 60, bullet = 30, laser = 60, energy = 60, bomb = 30, bio = 90, rad = 90) but good luck getting one of these.
* "Real wizard hats": Roughly on part with a HoS hat in terms of armor.

And on the mask/glasses front (ignoring the ones already mentioned):
* "Facehugger": If pun pun ends the round with a hugger on his face but doesn't get gibbed (very unlikely since he'd be infected at that point) he or his next of kin will start already impregnated. Obviously huge chance for shenanigans here, but everyone can tell he got hugged, and there's no way for anyone else to get that face hugger before it dies. If they let the larva walk free that's their own fault
*

**Comment by neersighted** [comment]
:+1: balance memes here are silly, if it becomes a problem we fix it.

**Comment by bgobandit** [comment]
Secret Technique

File: code/modules/mob/living/carbon/monkey/punpun.dm (hunk 1 above)
Comment (CosmicScientist) [inline code/modules/mob/living/carbon/monkey/punpun.dm]: George (Curious George, never saw it but I guess some people might like it).
Lucy (oldest known ancestor of humans unless that got changed at some point).
Darwin (Charles Darwin, so doing what the Wild Thornberries and I assume everyone else has done).

Comment (RemieRichards) [inline code/modules/mob/living/carbon/monkey/punpun.dm]: Aldo (Planet of the apes)
Caeser (Rise of the Planet of the apes)
Kong (King Kong/Donkey)
Professor Bobo (Mystery science theatre 3000)
Terk (Tarzan)

Comment (KorPhaeron) [inline code/modules/mob/living/carbon/monkey/punpun.dm]: I would rather have more teegee culture than random movie names personally but it's up to you Incoming

Comment (RemieRichards) [inline code/modules/mob/living/carbon/monkey/punpun.dm]: @KorPhaeron SS13 is full of -all- pop culture type references, soo, eat a dick? :3

Comment (KorPhaeron) [inline code/modules/mob/living/carbon/monkey/punpun.dm]: Yeah I know and I'm one of the worst offenders in adding them.

The monkeys traditionally have been jokes from traditional games though

**Comment by Shadowlight213** [comment]
Alright. thanks for the research. I'm fine with this.

**Comment by Lumipharon** [comment]
>reliable way to carry aliens to the next round

Oh lawdy.

File: code/modules/mob/living/carbon/monkey/punpun.dm (hunk 1 above)
Comment (KazeEspada) [inline code/modules/mob/living/carbon/monkey/punpun.dm]: Ominisificer(the 3.5 build purposely built to find and stop Pun Pun).

Chuck E. Cheese(3.5 build to break speed of light).

Chuck the farmer(3.5 chicken infested farmer to THROW FLAMING CHICKENS OF DEATH)

**Comment by YotaXP** [comment]
So will this clone Lamaar?

**Comment by Incoming5643** [comment]
It will clone lamaar (petception) 

It won't clone drones though, because it only carries over the type, which in this case would be an empty hat holder. Basically it'd be a drone hat that could never uncurl.

**Comment by MrPerson** [comment]
I should probably mention that this is a very limited take on what you can do with savefiles. Not only could you save the actual objects themselves instead of just their type, you could also save an entire datum and load the hat and mask from that. Fuck, you could save and load the entire fucking monkey if you wanted.

Not that I'm saying you should, obviously, I'm just saying you could. There's some resolvable issues with saving and loading things that I'll go over if you want. Saving vv'd shit is obviously a HUGE balance concenr. There's the possibility for subtle bugs like saving an object with invalid references. And of course there's a crazy amount of work to make that viable at all. Obviously what you're doing here is sufficient.

Whole thing seems ok to me. I think you should blacklist facehuggers though. Anything else is fine and probably not worth the effort to attempt to cheese, but people are going to try to sneak huggers into the next round even if that means not enjoying the current one as much.

**Comment by Incoming5643** [comment]
Fun fact when I was doing the last poly savefile work I had a var set up for poly to remember the person who killed him the last round for vengeance purposes. What I didn't realize was the var I was using to track the killer (I think it was last_attacker or something) wasn't just a name, but in fact the whole human mob.

So poly would commit the mob of his killer to his save file, then come the next round the killer was trapped in nullspace before the round even started in his last round body, referenced only by the bird's save vars.

For the record a spessmen has about a 2.5 mb size as a savefile.

**Comment by Iamgoofball** [comment]
Keep in mind facehuggers will implant the mob and since about 3-5 people spawn staring at Pun-Pun thats a high chance for them to immediately stuff the monkey into the gibber if he's got a facehugger on him.

**Comment by Zilenan91** [comment]
That's why it's fine

**Comment by Lumipharon** [comment]
Dude, people go out of their way to jump into a cage with a live alien queen or face hugger in it.

I seriously think you under estimate the playerbase if you don't think that given the chance, plenty of people would be happy to nurture xenopun.

**Comment by KazeEspada** [comment]
LumiIHATEFUNpharon everyone.

**Comment by Cheridan** [comment]
BREAK THE CYCLE OF ETERNAL RETURN

**Comment by Incoming5643** [comment]
There was gonna be more monkey names and a map edit but whatever I'll make a new pull

**Pull request merged**